### DISCUSSION TITLE
useKeyPress指定target时如何支持更多event?

### DISCUSSION THREAD

**Discussion by xmsz** [alibaba/hooks#1878] [Q&A]
背景
- 我想在div.a里触发回调
- 但是发现触发不了
- 必须操作div.a里的input才能触发？

还是因为我是在弹窗(dialog)里触发
我发现useFocusWithin也不行，但是useHover可以

**Comment by SignDawn** [comment]
这不提供一个代码示例？

**Reply by xmsz** [reply]
https://codesandbox.io/s/chuan-ru-dom-yuan-su-forked-o33ym9?file=/App.tsx

**Reply by SignDawn** [reply]
我看了下源码，大概是在 Dialog 组件中，这个 div dom 元素初始化的时候没有被创建。
![image](https://user-images.githubusercontent.com/48878568/190838784-0647ac70-97e0-4f11-a4e6-e73ffb2dc587.png)

**Reply by xmsz** [reply]
> 我看了下源码，大概是在 Dialog 组件中，这个 div dom 元素初始化的时候没有被创建。 ![image](https://user-images.githubusercontent.com/48878568/190838784-0647ac70-97e0-4f11-a4e6-e73ffb2dc587.png)

那这个有什么解决/规避办法吗  还是这个是一个fusion.Dialog的bug

**Reply by SignDawn** [reply]
看官方怎么解决吧，我提了一个 issue 。
你这个 Dialog ，默认必须显示吗？我试了初始的时候不显示的情况，是可以的。
`/**
 * title: Pass in DOM element
 * desc: Pass in a function that returns the DOM element.
 *
 * title.zh-CN: 传入 DOM 元素
 * desc.zh-CN: 传入 function 并返回一个 dom 元素。
 */

import { Dialog } from '@alifd/next';
import { useFocusWithin } from 'ahooks';
import React, { useRef, useState } from 'react';
export default () => {
  const ref = useRef(null);
  const isFocusWithin = useFocusWithin(ref);

  const [visible, setVisible] = useState(false);

  return (
    <>
      <Dialog
        visible={visible}
        overlayProps={{
          cache: true,
        }}
        onClose={() => setVisible(false)}
      >
        <div
          ref={ref}
          style={{
            padding: 16,
            backgroundColor: isFocusWithin ? 'red' : '',
            border: '1px solid gray',
          }}
        >
          <label style={{ display: 'block' }}>
            First Name: <input />
          </label>
          <label style={{ display: 'block', marginTop: 16 }}>
            Last Name: <input />
          </label>
        </div>
        <p>isFocusWithin: {JSON.stringify(isFocusWithin)}</p>
      </Dialog>
      <button onClick={() => setVisible(true)}>??</button>
    </>
  );
};
`

**Reply by xmsz** [reply]
> 你这个 Dialog ，默认必须显示吗？我试了初始的时候不显示的情况，是可以的。

不必须显示

但是，即使切换，也只能点击input才能触发，其他区域触发不了
https://codesandbox.io/s/chuan-ru-dom-yuan-su-forked-yp2enl?file=/App.tsx

**Reply by SignDawn** [reply]
什么不能触发？你希望什么能够触发？ a 标签？改下例子吧

**Reply by xmsz** [reply]
稍等我修改一下

**Reply by xmsz** [reply]
> 什么不能触发？你希望什么能够触发？ a 标签？改下例子吧

我忘记浏览器对于非可编辑组件不能触发keydown这个限制了 https://segmentfault.com/a/1190000041568137 

我的需求是
- 当用户鼠标在某个dom里，点击某个按键时触发

然后一开始用keydown，没有绑定target，正常，但是鼠标在整个位置都会触发
然后我就把target改成了dom，结果没有效果，以为是我绑定有问题


我现在用useKeypress +  useHover来解决需求了
